file/shard.go

```diff
@@ -0,0 +1,151 @@
+package file
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	dagpb "github.com/ipld/go-codec-dagpb"
+	"github.com/ipld/go-ipld-prime"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/node/basicnode"
+	"github.com/multiformats/go-multicodec"
+)
+
+type shardNodeFile struct {
+	ctx       context.Context
+	lsys      *ipld.LinkSystem
+	substrate ipld.Node
+	done      bool
+	rdr       io.Reader
+}
+
+var _ ipld.Node = (*shardNodeFile)(nil)
+
+func (s *shardNodeFile) Read(p []byte) (int, error) {
+	if s.done {
+		return 0, io.EOF
+	}
+	// collect the sub-nodes on first use
+	if s.rdr == nil {
+		links, err := s.substrate.LookupByString("Links")
+		if err != nil {
+			return 0, err
+		}
+		readers := make([]io.Reader, 0)
+		lnki := links.ListIterator()
+		for !lnki.Done() {
+			_, lnk, err := lnki.Next()
+			if err != nil {
+				return 0, err
+			}
+			if pbl, ok := lnk.(dagpb.PBLink); ok {
```

Comment:
hm, why do we use LookupByString("Links") above, and then a type assertion directly on a dagpb type here? seems like we should consistently be flexible to allow any IPLD node satisfying the same shape, or consistently assume dagpb types for simplicity/speed. Probably the former, to avoid unnecessary ties on go-codec-dagpb.